ci-operator/step-registry/azure/provision/service-principal/minimal-permission/azure-provision-service-principal-minimal-permission-commands.sh

```diff
@@ -136,282 +104,39 @@ fi
-\"Microsoft.Storage/register/action\",
-\"Microsoft.ManagedIdentity/register/action\"
-"""
-    sts_role_name="${CLUSTER_NAME}-custom-role-sts"
-    sts_role_definition="${ARTIFACT_DIR}/azure-custom-role-definition-sts-minimal-permissions.json"
-    sts_sp_name="${CLUSTER_NAME}-sp-sts"
-    sts_sp_output=$(mktemp)
-
-    echo "Create SP with minimal permission for CCO"
-    create_role_definition_json "${sts_role_name}" "${sts_required_permissions}" "${sts_role_definition}"
-    create_custom_role "${sts_role_definition}" "${sts_role_name}"
-    # for destroy
-    custom_role_name_json=$(echo "${custom_role_name_json}" | jq -c -S ". +={\"ccoctl\":\"${sts_role_name}\"}")
-    echo "${custom_role_name_json}" > "${SHARED_DIR}/azure_custom_role_name"
-    sts_sp_role="${sts_role_name}"
-
-    create_sp_with_custom_role "${sts_sp_name}" "${sts_sp_role}" "${AZURE_AUTH_SUBSCRIPTOIN_ID}" "${sts_sp_output}"
-    sts_sp_id=$(jq -r .appId "${sts_sp_output}")
-    sts_sp_password=$(jq -r .password "${sts_sp_output}")
-    sts_sp_tenant=$(jq -r .tenant "${sts_sp_output}")
+for sp_type in ${AZURE_MINIMAL_PERMISSIONS_USAGE}; do
+    role_name=$(jq -r ".${sp_type}" "${SHARED_DIR}/azure_custom_role_name")
+    sp_name="${CLUSTER_NAME}-sp-${sp_type}"
+    sp_output="$(mktemp)"
+    if [[ -n "${AZURE_PERMISSION_FOR_CLUSTER_SP}" ]] && [[ "${role_name}" == "cluster" ]]; then
+        role_name="${AZURE_PERMISSION_FOR_CLUSTER_SP}"
+    fi
 
-    if [[ "${sts_sp_id}" == "" ]] || [[ "${sts_sp_password}" == "" ]]; then
+    echo "Creating ${sp_type} sp with role ${role_name} granted..."
+    create_sp_with_custom_role "${sp_name}" "${role_name}" "${AZURE_AUTH_SUBSCRIPTOIN_ID}" "${sp_output}"    
+    sp_id=$(jq -r .appId "${sp_output}")
+    sp_password=$(jq -r .password "${sp_output}")
+    sp_tenant=$(jq -r .tenant "${sp_output}")
+    if [[ "${sp_id}" == "" ]] || [[ "${sp_password}" == "" ]]; then
         echo "Unable to get service principal id or password, exit..."
         exit 1
     fi
 
-    echo "New service principal id: ${sts_sp_id}"
-    cat <<EOF > "${SHARED_DIR}/azure_minimal_permission_sts"
-{"subscriptionId":"${AZURE_AUTH_SUBSCRIPTOIN_ID}","clientId":"${sts_sp_id}","tenantId":"${sts_sp_tenant}","clientSecret":"${sts_sp_password}"}
+    echo "New service principal id: ${sp_id}"
+    os_sp_file_name="azure_minimal_permission"
+    if [[ "${role_name}" != "cluster" ]]; then
+        os_sp_file_name="azure_minimal_permission_${sp_type}"
+    fi
+    cat <<EOF > "${SHARED_DIR}/${os_sp_file_name}"
+{"subscriptionId":"${AZURE_AUTH_SUBSCRIPTOIN_ID}","clientId":"${sp_id}","tenantId":"${sp_tenant}","clientSecret":"${sp_password}"}
 EOF
+
     # for destroy
-    echo "${sts_sp_id}" >> "${SHARED_DIR}/azure_sp_id"
-    rm -rf ${sts_sp_output}
-fi
+    echo "${sp_id}" > "${SHARED_DIR}/azure_sp_id"
+    rm -f ${sp_output}
+
+    # ensure that role assignment creation is successful
+    cmd="az role assignment list --role {role_name}"
```

Comment:
This step is used to create sp, no actions related to `az role assignment`, do we need to check it here ?

Reply:
Here create sp with customer role, and need sometime to waiting for role assignment successful, else following steps might fail, e.g step `upi-install-azure`, so add this check to ensure that role assignment succeeded.